BOLT 12 static invoice encoding and building

valentinewallace opened the pull request:
Defines an interface for static BOLT 12 invoices, based on https://github.com/lightning/bolts/pull/1149. 

Similar to single-use BOLT 12 invoices, except without a payment hash or amount. Currently may only be created from `Offer`s.

valentinewallace reviewed (commented):
Conceptual feedback appreciated. Left some comments calling out specific things that might warrant discussion :)

valentinewallace commented:
Pushed a fix for the test.

valentinewallace commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,591 @@
+	relative_expiry: Option<Duration>,
+	fallbacks: Option<Vec<FallbackAddress>>,
+	features: Bolt12InvoiceFeatures,
+	signing_pubkey: PublicKey,
```
These contents are missing the invoice message paths to send the `held_htlc_available` message over, will fix in the next push.

jkczyz commented:
> Haven't taken a thorough look yet, but yeah my feeling was to make a separate type for static invoices. Maybe there could be a separate `InvoiceContents` variant instead, but it wouldn't be able to reuse `InvoiceFields` since that requires a `PaymentHash`. And then `Invoice::payment_hash` would need to return an `Option`, which I don't think we want.

A counterpoint might be that with https://github.com/lightningdevkit/rust-lightning/pull/3078 we might want to have a single invoice type for the event. In that case, maybe we can live with an `Option` and where `send_payment_for_bolt12_invoice` can pay either. @valentinewallace What will the static invoice handling code look like? Will the onion message use a separate TLV for such invoice? Or will it reuse `66`? Just thinking about the implications when it comes to parsing.

valentinewallace commented:
For reference, here are the pros and cons of using a separate type from where I stand:

Pros of separate type:
- Avoids `Option`al return value in `payment_hash` and `amount_msat` methods 
- Avoids having non-applicable `payer_*` methods on static invoices, i.e. payer_id, invreq features
- Because static invoices are required to come from offers (not refunds): 
    - their docs can be more custom to them and not reference refunds
    - methods like `offer_features` can return non-Options 

Cons of separate type:
- a lot of not-easily-DRY’d extra code
- may be annoying if we want to include static invoices in e.g. `InvoiceReceived` events
- `PaymentPurpose` looks like it might need some reworking for static invoices (though this may also apply to re-purposing `Bolt12Invoice`)

I may be missing some since I didn’t go too far down the same-type road but this is what’s come up so far.

valentinewallace commented:
> What will the static invoice handling code look like? Will the onion message use a separate TLV for such invoice? Or will it reuse 66? Just thinking about the implications when it comes to parsing.

Feel free to check out `handle_message` in the top commit of my WIP branch for handling: https://github.com/valentinewallace/rust-lightning/tree/2024-04-bolt12-keysend-invoice-2 (it doesn't include the async flow at all, just pays the static invoice immediately on receipt). 

I used a separate type for static invoices, not 66.

valentinewallace commented:
> Isn't part of the point of #3078 that people might want to look at the payment hash before they pay, in which case a single type won't help?

Is it possible that users may want to set their own keysend preimage? Not sure if `InvoiceReceived` may still be useful for them.

jkczyz commented:
> Isn't part of the point of #3078 that people might want to look at the payment hash before they pay, in which case a single type won't help?

Respond there regarding the sending case. It's more for receiving when the payment hash is important.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1130 @@
+	}
+}
+
+tlv_stream!(InvoiceTlvStream, InvoiceTlvStreamRef, 160..240, {
+	(160, payment_paths: (Vec<BlindedPath>, WithoutLength, Iterable<'a, BlindedPathIter<'a>, BlindedPath>)),
+	(162, blindedpay: (Vec<BlindedPayInfo>, WithoutLength, Iterable<'a, BlindedPayInfoIter<'a>, BlindedPayInfo>)),
+	(164, created_at: (u64, HighZeroBytesDroppedBigSize)),
+	(166, relative_expiry: (u32, HighZeroBytesDroppedBigSize)),
+	(172, fallbacks: (Vec<FallbackAddress>, WithoutLength)),
+	(174, features: (Bolt12InvoiceFeatures, WithoutLength)),
+	(176, node_id: PublicKey),
+	(178, message_paths: (Vec<BlindedPath>, WithoutLength)),
+});
```
Should we re-use the TLV stream from `invoice.rs`? That was the approach that we used for `Refund` even though the fields have different requirements.

valentinewallace replied:
Hmm, my reasoning was that this is <10 lines and leans on the type system whereas manual checks for `payment_hash`, amount, and message paths would be more verbose/error prone and require new `Bolt12SemanticError` variant(s). Do you think it's better to reuse it?

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,585 @@
+		if invoice.payment_paths.is_empty() {
+			return Err(Bolt12SemanticError::MissingPaths);
+		}
+		if invoice.offer.chains().len() > 1 {
```
Should this be checked on the passed `Offer` prior to creating the `InvoiceContents`

valentinewallace replied:
True, I went ahead and fixed that in the latest push.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,585 @@
+	/// after `created_at`.
+	pub fn for_offer_using_keys(
+		offer: &'a Offer, payment_paths: Vec<(BlindedPayInfo, BlindedPath)>, created_at: Duration,
+		keys: KeyPair,
```
I think we we can make the `static_invoice` mod private in the meantime.

valentinewallace replied:
Made the module `pub(crate)`!

valentinewallace commented:
Added tests and fixed up some things in the latest push so gonna go ahead and take this out of draft. Still a few outstanding TODOs in the PR description. Thanks @jkczyz for the review 🫡

valentinewallace commented:
Addressed feedback, this should be good for review. 

Going to skip adding code example module-level docs until the methods to create these new structs exist on `ChannelManager`, since that's currently the intended usage. 

Looks like this needs rebase as well so lmk if I can go ahead with that.

jkczyz commented:
> Looks like this needs rebase as well so lmk if I can go ahead with that.

Yes, please go ahead and rebase.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,359 @@
+
+		let features = features.unwrap_or_else(Bolt12InvoiceFeatures::empty);
+
+		let signing_pubkey = match node_id {
+			None => return Err(Bolt12SemanticError::MissingSigningPubkey),
+			Some(node_id) => {
+				let offer_node_id =
+					offer_tlv_stream.node_id.ok_or(Bolt12SemanticError::MissingSigningPubkey)?;
+				if node_id != offer_node_id {
+					return Err(Bolt12SemanticError::InvalidSigningPubkey);
+				}
+				node_id
+			},
+		};
```
Looks like this will need to account for the recent spec change that allows leaving out `offer_tlv_stream.node_id` if a blinded path is included.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -72,6 +74,88 @@ struct InvoiceContents {
+pub struct StaticInvoiceBuilder<'a> {
+	offer_bytes: &'a Vec<u8>,
+	invoice: InvoiceContents,
+	keys: KeyPair,
```
Could we derive the keys from the offer metadata rather than requiring that the caller pass them?

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -158,6 +300,53 @@ impl StaticInvoice {
+		}
+	}
+
+	fn as_invoice_fields_tlv_stream(&self) -> InvoiceTlvStreamRef {
```
Could we include the offer TLV stream here, too, for consistency with the other code? (i.e., return a `PartialInvoiceTlvStreamRef`). I realize it won't be used later but it is nice for testing.

valentinewallace commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,359 @@
+impl TryFrom<(OfferTlvStream, InvoiceTlvStream)> for InvoiceContents {
+	type Error = Bolt12SemanticError;
+
+	fn try_from(tlv_stream: (OfferTlvStream, InvoiceTlvStream)) -> Result<Self, Self::Error> {
```
I find the indirection of the wrapper type less readable, tbh. "Partial" sounds pretty vague to me.

jkczyz commented on lightning/src/offers/invoice.rs:
```diff
@@ -1299,10 +1302,12 @@ impl TryFrom<PartialInvoiceTlvStream> for InvoiceContents {
 			},
 		) = tlv_stream;
 
+		if message_paths.is_some() { return Err(Bolt12SemanticError::UnexpectedPaths) }
```
Could you update the parsing tests?

valentinewallace commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1131 @@
+		}
+
+		// Error if offer paths are missing.
+		let mut offer_without_paths = valid_offer.clone();
+		let mut offer_tlv_stream = offer_without_paths.as_tlv_stream();
+		offer_tlv_stream.paths.take();
+		let mut buffer = Vec::new();
+		offer_tlv_stream.write(&mut buffer).unwrap();
+		offer_without_paths = Offer::try_from(buffer).unwrap();
```
IIRC I needed to build it with paths originally to force a derived signing pubkey, does that jibe with your understanding?

valentinewallace replied:
Nice for the test to be resilient to check reordering. Lmk if you feel strongly about this.

valentinewallace commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1131 @@
+		assert!(invoice.fallbacks().is_empty());
+		assert_eq!(invoice.invoice_features(), &Bolt12InvoiceFeatures::empty());
+
+		let keys_opt = offer.verify(&expanded_key, &secp_ctx).unwrap().1;
```
Added 353dd61cb05c89dbb626da83c32913f8ba76957d, lmk if that's not what you meant.

valentinewallace commented:
Addressed feedback.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1173 @@
+	///
+	/// Unless [`StaticInvoiceBuilder::relative_expiry`] is set, the invoice will expire 24 hours
+	/// after `created_at`.
+	pub fn for_offer_using_keys<T: secp256k1::Signing>(
```
Could you name this `for_offer_using_derived_keys`. Given this now takes `ExpandedKey`, that naming is better suited than when it took `KeyPair` and is more consistent with similar API.

Alternatively, add a method to `Offer` using the "derived" naming taking `ExpandedKey` and revert this back to `KeyPair` to align better with `InvoiceRequest` and `InvoiceBuilder`.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1173 @@
+}
+
+/// A semantically valid [`StaticInvoice`] that hasn't been signed.
+pub struct UnsignedStaticInvoice {
```
Looks like this only needs to be `pub` if exposed in an `InvoiceBuilder::build` method. Should we go ahead and add a version that works with an `Offer` not using derived keys? Could leave as a follow-up, too.

valentinewallace replied:
I'm good to add that in follow-up!

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1173 @@
+			&expanded_key,
+			&secp_ctx,
+		)
+		.unwrap()
+		.build_and_sign(&secp_ctx)
+		.unwrap();
```
Indent? Likewise throughout.

valentinewallace replied:
I'm using rustfmt 😀

jkczyz replied:
Huh, strange that it's inconsistent with when the first call can fit on one line.

valentinewallace commented:
Addressed feedback. Lmk when I can squash, @jkczyz.

jkczyz commented:
> Addressed feedback. Lmk when I can squash, @jkczyz.

Yes, please squash

valentinewallace commented:
Squashed with no changes.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1173 @@
+		$contents.issuer()
+	}
+
+	/// Paths to the node that may supply the invoice on the recipient's behalf, originating from
+	/// publicly reachable nodes. Taken from [`Offer::paths`].
+	///
+	/// [`Offer::paths`]: crate::offers::offer::Offer::paths
+	pub fn offer_message_paths(&$self) -> &[BlindedPath] {
+		$contents.offer_message_paths()
+	}
+
+	/// Paths to the recipient for indicating that a held HTLC is available to claim when they next
+	/// come online.
+	pub fn message_paths(&$self) -> &[BlindedPath] {
+		$contents.message_paths()
+	}
```
It's a bit inconsistent that `Bolt12Invoice::message_paths` and `StaticInvoice::message_paths` don't correspond to the same thing. We may want to re-think the naming here. Further, `message_paths` may not be descriptive enough especially when we also add paths for payment notification to the offer. Though I'm having trouble finding a way to differentiate naming the latter from the one added here.

Possible naming:
- `request_paths` for `offer_paths` TLV
- `payment_paths` for `invoice_paths` and `invoice_blindedpay`
- `notification_paths` for paths add to `offer` for upcoming payment notifications
- ??? for paths added to `invoice` for async payments

WDYT?

valentinewallace replied:
Don't feel super strongly since users are unlikely to be accessing these fields directly, but I'm happy to rename these methods. Since we're pulling in other considerations like payment notifications and renaming offer fields as well, I'm working on a follow-up based on this PR that should be open sometime today, if that works for you. Naming suggestions SGTM.

valentinewallace replied:
Opened #3118.

jkczyz commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1173 @@
+		$contents.offer_message_paths()
+	}
+
+	/// Paths to the recipient for indicating that a held HTLC is available to claim when they next
```
For refreshing my understanding, who is holding the HTLC here? Is this for sending `held_htlc_available`? Do you have a reference for this flow? I didn't see it in the spec PR other than:

>   - if `invoice_payment_hash` is unset:
>     - MUST reject the invoice if `invoice_message_paths` is not present or is empty.
>     - MUST pay asynchronously using the `held_htlc_available` onion message
>       flow, where the onion message is sent over `invoice_message_paths`.

valentinewallace replied:
> who is holding the HTLC here?

Either the always-online sender or the often-offline sender's LSP. 

> Is this for sending held_htlc_available?

Yep!

For reference, https://github.com/lightning/bolts/pull/1149 is based on https://github.com/lightning/bolts/pull/989, which is the flow referred to in the quote. Let me know if something's unclear.

jkczyz replied:
Ah, thanks! I was only looking at the last commits and not also the first ones. 🤦‍♂️

valentinewallace commented on lightning/src/offers/invoice.rs:
```diff
@@ -135,7 +135,7 @@ use std::time::SystemTime;
 pub(crate) const DEFAULT_RELATIVE_EXPIRY: Duration = Duration::from_secs(7200);
 
-/// Tag for the hash function used when signing a [`Bolt12Invoice`]'s merkle root.
+/// Tag for the hash function used when signing a BOLT 12 invoice's merkle root.
```
This aligns with the current spec, since static invoices are still officially "invoice" messages. We could add a special exception for static invoices, though, no strong opinion.

valentinewallace commented:
Went ahead and squashed in updates:

```diff
diff --git a/lightning/src/offers/invoice.rs b/lightning/src/offers/invoice.rs
index c570aed66..6695121b3 100644
--- a/lightning/src/offers/invoice.rs
+++ b/lightning/src/offers/invoice.rs
@@ -135,7 +135,7 @@ use std::time::SystemTime;

 pub(crate) const DEFAULT_RELATIVE_EXPIRY: Duration = Duration::from_secs(7200);

-/// Tag for the hash function used when signing a BOLT 12 invoice's merkle root.
+/// Tag for the hash function used when signing a [`Bolt12Invoice`]'s merkle root.
 pub const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice", "signature");

 /// Builds a [`Bolt12Invoice`] from either:
@@ -1164,7 +1164,7 @@ tlv_stream!(InvoiceTlvStream, InvoiceTlvStreamRef, 160..240, {
        (174, features: (Bolt12InvoiceFeatures, WithoutLength)),
        (176, node_id: PublicKey),
        // Only present in `StaticInvoice`s.
-       (178, message_paths: (Vec<BlindedPath>, WithoutLength)),
+       (238, message_paths: (Vec<BlindedPath>, WithoutLength)),
 });

 pub(super) type BlindedPathIter<'a> = core::iter::Map<
diff --git a/lightning/src/offers/static_invoice.rs b/lightning/src/offers/static_invoice.rs
index dd132bde7..77e2b1211 100644
--- a/lightning/src/offers/static_invoice.rs
+++ b/lightning/src/offers/static_invoice.rs
@@ -17,7 +17,6 @@ use crate::ln::msgs::DecodeError;
 use crate::offers::invoice::{
        check_invoice_signing_pubkey, construct_payment_paths, filter_fallbacks, BlindedPathIter,
        BlindedPayInfo, BlindedPayInfoIter, FallbackAddress, InvoiceTlvStream, InvoiceTlvStreamRef,
-       SIGNATURE_TAG,
 };
 use crate::offers::invoice_macros::{invoice_accessors_common, invoice_builder_methods_common};
 use crate::offers::merkle::{
@@ -43,8 +42,11 @@ use crate::offers::invoice::is_expired;
 #[allow(unused_imports)]
 use crate::prelude::*;

-/// Static invoices default to expiring after 24 hours.
-const DEFAULT_RELATIVE_EXPIRY: Duration = Duration::from_secs(3600 * 24);
+/// Static invoices default to expiring after 2 weeks.
+const DEFAULT_RELATIVE_EXPIRY: Duration = Duration::from_secs(3600 * 24 * 14);
+
+/// Tag for the hash function used when signing a [`StaticInvoice`]'s merkle root.
+pub const SIGNATURE_TAG: &'static str = concat!("lightning", "static_invoice", "signature");

 /// A `StaticInvoice` is a reusable payment request corresponding to an [`Offer`].
 ///
@@ -566,13 +568,13 @@ mod tests {
        use crate::ln::features::{Bolt12InvoiceFeatures, OfferFeatures};
        use crate::ln::inbound_payment::ExpandedKey;
        use crate::ln::msgs::DecodeError;
-       use crate::offers::invoice::{InvoiceTlvStreamRef, SIGNATURE_TAG};
+       use crate::offers::invoice::InvoiceTlvStreamRef;
        use crate::offers::merkle;
        use crate::offers::merkle::{SignatureTlvStreamRef, TaggedHash};
        use crate::offers::offer::{Offer, OfferBuilder, OfferTlvStreamRef, Quantity};
        use crate::offers::parse::{Bolt12ParseError, Bolt12SemanticError};
        use crate::offers::static_invoice::{
-               StaticInvoice, StaticInvoiceBuilder, DEFAULT_RELATIVE_EXPIRY,
+               StaticInvoice, StaticInvoiceBuilder, DEFAULT_RELATIVE_EXPIRY, SIGNATURE_TAG,
        };
        use crate::offers::test_utils::*;
        use crate::sign::KeyMaterial;
```

valentinewallace commented:
Squashed in the latest change with the following diff:

```diff
diff --git a/lightning/src/offers/static_invoice.rs b/lightning/src/offers/static_invoice.rs
index 77e2b1211..c6b6fa766 100644
--- a/lightning/src/offers/static_invoice.rs
+++ b/lightning/src/offers/static_invoice.rs
@@ -428,8 +428,7 @@ impl InvoiceContents {
        }

        fn fallbacks(&self) -> Vec<Address> {
-               let chain =
-                       self.offer.chains().first().cloned().unwrap_or_else(|| self.offer.implied_chain());
+               let chain = self.chain();
                self.fallbacks
                        .as_ref()
                        .map(|fallbacks| filter_fallbacks(chain, fallbacks))

valentinewallace commented on lightning/src/offers/static_invoice.rs:
```diff
@@ -0,0 +1,1174 @@
+
+/// Builds a [`StaticInvoice`] from an [`Offer`].
+///
+/// See [module-level documentation] for usage.
```
Oops, I forgot to update these docs when deciding to punt module-level docs. Updated.

valentinewallace commented:
Updated the outdated docs:

```diff
diff --git a/lightning/src/offers/static_invoice.rs b/lightning/src/offers/static_invoice.rs
index c6b6fa766..d0846b29a 100644
--- a/lightning/src/offers/static_invoice.rs
+++ b/lightning/src/offers/static_invoice.rs
@@ -82,12 +82,9 @@ struct InvoiceContents {

 /// Builds a [`StaticInvoice`] from an [`Offer`].
 ///
-/// See [module-level documentation] for usage.
-///
 /// [`Offer`]: crate::offers::offer::Offer
-/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
-/// [`Bolt12Invoice`]: crate::offers::invoice::Bolt12Invoice
 /// This is not exported to bindings users as builder patterns don't map outside of move semantics.
+// TODO: add module-level docs and link here
 pub struct StaticInvoiceBuilder<'a> {
        offer_bytes: &'a Vec<u8>,
        invoice: InvoiceContents,

valentinewallace commented:
> Please add fuzzing in a followup :)

Will do, I added this to the tracking issue and will hopefully get to it soon.
